File: mkr-android-lib/src/main/java/com/mkrworld/androidlib/ui/adapter/BaseAdapterItem.java
package com.mkrworld.androidlib.ui.adapter;

public class BaseAdapterItem<MKR> {
    private int mViewType;
    private MKR mMkr;

    /**
     * Constructor
     *
     * @param viewType
     * @param d         Object Data pass to the BaseViewHolder
     */
    public BaseAdapterItem(int viewType, MKR d) {
        mViewType = viewType;
        mMkr = d;
    }

    public MKR getBindingData() {
        return mMkr;
    }

    public int getmViewType() {
        return mViewType;
    }

    @Override
    public boolean equals(Object obj) {
        try {
            if (obj instanceof BaseAdapterItem) {
                BaseAdapterItem baseAdapterItem = (BaseAdapterItem) obj;
                if (baseAdapterItem.mViewType == mViewType && baseAdapterItem.mMkr.equals(mMkr)) {
                    return true;
                }
            }
        } catch (Exception e) {
            // Any exception occur at the time of equals operation
        }
        return false;
    }

    @Override
    public int hashCode() {
        return mMkr.hashCode();
    }
}

File: mkr-android-lib/src/main/java/com/mkrworld/androidlib/ui/adapter/BaseAdapterItem.kt
package com.mkrworld.androidlib.ui.adapter

class BaseAdapterItem<MKR> {
    private var mViewType : Int? = null
    private var mMkr : MKR? = null

    /**
     * Constructor
     *
     * @param viewType
     * @param mkr         Object Data pass to the BaseViewHolder
     */
    constructor(viewType : Int, mkr : MKR) {
        mViewType = viewType
        mMkr = mkr
    }

    fun getViewType() : Int {
        return mViewType!!
    }

    fun getData() : MKR{
        return mMkr!!
    }

    override fun equals(obj : Any?) : Boolean {
        try {
            if (obj is BaseAdapterItem<*>) {
                val adapterItem = obj as BaseAdapterItem<*>?
                if (adapterItem !!.mViewType == mViewType && adapterItem.mMkr == mMkr) {
                    return true
                }
            }
        } catch (e : Exception) {
            // Any exception occur at the time of equals operation
        }

        return false
    }

    override fun hashCode() : Int {
        return mMkr!!.hashCode()
    }
}
